src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -1584,15 +2088,293 @@ tx_gateway_frontend::do_commit_tm_tx(
     try {
         if (
           tx.status != tm_transaction::tx_status::ongoing
-          && tx.status != tm_transaction::tx_status::preparing) {
+          && tx.status != tm_transaction::tx_status::preparing
+          && tx.status != tm_transaction::tx_status::prepared
+          && tx.status != tm_transaction::tx_status::aborting) {
+            vlog(
+              txlog.warn,
+              "can't commit {} with pid:{} unexpected status:{}",
+              tx.id,
+              tx.pid,
+              tx.status);
             outcome->set_value(tx_errc::invalid_txn_state);
             co_return tx_errc::invalid_txn_state;
         }
 
         if (tx.status == tm_transaction::tx_status::ongoing) {
             if (tx.etag != expected_term) {
-                outcome->set_value(tx_errc::invalid_txn_state);
-                co_return tx_errc::invalid_txn_state;
+                vlog(
+                  txlog.trace,
+                  "commit encountered old ongoing tx:{} etag:{} pid:{} "
+                  "tx_seq:{} in term:{}, fetching state from old leader",
+                  tx.id,
+                  tx.etag,
+                  tx.pid,
+                  tx.tx_seq,
+                  expected_term);
+                auto r1 = co_await dispatch_fetch_tx(tx.id, tx.etag);
+                if (!r1) {
+                    vlog(
+                      txlog.trace,
+                      "can't find old tx:{} etag:{} status:{} term:{}",
+                      tx.id,
+                      tx.etag,
+                      tx.status,
+                      expected_term);
+                    outcome->set_value(tx_errc::unknown_server_error);
```

Comment:
I already have a catch at the top level (end_txn) but just in case added here too

Reply:
> but just in case added here too

i don't think there is any need to not trust try {} block to do the right thing, i just didn't see it.

Reply:
`fetch_tx` doesn't throw exceptions but just in case we have several safety checks

 - https://github.com/rystsov/redpanda/blob/b154b1b3e0efe6d77c7413a65c04a72fabf61b37/src/v/cluster/tx_gateway_frontend.cc#L335
 - https://github.com/rystsov/redpanda/blob/b154b1b3e0efe6d77c7413a65c04a72fabf61b37/src/v/cluster/tx_gateway_frontend.cc#L1555
 - https://github.com/rystsov/redpanda/blob/b154b1b3e0efe6d77c7413a65c04a72fabf61b37/src/v/cluster/tx_gateway_frontend.cc#L1678